File: src/typescript/src/infrastructure/mappers/LampMapper.test.ts
import { LampMapper } from './LampMapper.ts';
import type { LampEntity } from '../../domain/entities/LampEntity.ts';

describe('LampMapper', () => {
  describe('toApiModel', () => {
    it('should convert domain entity to API model', () => {
      const entity: LampEntity = {
        id: '123',
        status: true,
        createdAt: '2025-01-01T00:00:00.000Z',
        updatedAt: '2025-01-01T00:00:00.000Z',
      };

      const apiModel = LampMapper.toApiModel(entity);

      expect(apiModel).toEqual({
        id: '123',
        status: true,
        createdAt: '2025-01-01T00:00:00.000Z',
        updatedAt: '2025-01-01T00:00:00.000Z',
      });
    });
  });

  describe('toDomainEntity', () => {
    it('should convert API model to domain entity', () => {
      const apiModel = {
        id: '456',
        status: false,
        createdAt: '2025-01-02T00:00:00.000Z',
        updatedAt: '2025-01-02T00:00:00.000Z',
      };

      const entity = LampMapper.toDomainEntity(apiModel);

      expect(entity).toEqual({
        id: '456',
        status: false,
        createdAt: '2025-01-02T00:00:00.000Z',
        updatedAt: '2025-01-02T00:00:00.000Z',
      });
    });
  });

  describe('toDomainEntityCreate', () => {
    it('should convert API create model to domain create model', () => {
      const apiModel = { status: true };

      const domainModel = LampMapper.toDomainEntityCreate(apiModel);

      expect(domainModel).toEqual({ status: true });
    });
  });

  describe('toDomainEntityUpdate', () => {
    it('should convert API update model to domain update model', () => {
      const apiModel = { status: false };

      const domainModel = LampMapper.toDomainEntityUpdate(apiModel);

      expect(domainModel).toEqual({ status: false });
    });
  });
});

File: src/kotlin/src/test/kotlin/com/lampcontrol/mapper/LampMapperTest.kt
package com.lampcontrol.mapper

import com.lampcontrol.api.models.Lamp
import com.lampcontrol.api.models.LampCreate
import com.lampcontrol.api.models.LampUpdate
import com.lampcontrol.entity.LampEntity
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Test
import java.time.Instant
import java.util.UUID

class LampMapperTest {
    private val mapper = LampMapper()

    @Test
    fun `should convert domain entity to API model`() {
        // Given
        val uuid = UUID.randomUUID()
        val now = Instant.now()
        val entity =
            LampEntity(
                id = uuid,
                status = true,
                createdAt = now,
                updatedAt = now,
            )

        // When
        val apiModel = mapper.toApiModel(entity)

        // Then
        assertEquals(uuid, apiModel.id)
        assertEquals(true, apiModel.status)
        assertEquals(now.toString(), apiModel.createdAt)
        assertEquals(now.toString(), apiModel.updatedAt)
    }

    @Test
    fun `should convert API model to domain entity`() {
        // Given
        val uuid = UUID.randomUUID()
        val timestamp = "2023-01-01T00:00:00Z"
        val apiModel =
            Lamp(
                id = uuid,
                status = false,
                createdAt = timestamp,
                updatedAt = timestamp,
            )

        // When
        val entity = mapper.toDomainEntity(apiModel)

        // Then
        assertEquals(uuid, entity.id)
        assertEquals(false, entity.status)
        assertEquals(Instant.parse(timestamp), entity.createdAt)
        assertEquals(Instant.parse(timestamp), entity.updatedAt)
    }

    @Test
    fun `should create domain entity from API create model`() {
        // Given
        val lampCreate = LampCreate(status = true)

        // When
        val entity = mapper.toDomainEntityCreate(lampCreate)

        // Then
        assertNotNull(entity.id)
        assertEquals(true, entity.status)
        assertNotNull(entity.createdAt)
        assertNotNull(entity.updatedAt)
    }

    @Test
    fun `should update domain entity from API update model`() {
        // Given
        val originalEntity = LampEntity.create(false)
        val lampUpdate = LampUpdate(status = true)

        // When
        val updatedEntity = mapper.updateDomainEntity(originalEntity, lampUpdate)

        // Then
        assertEquals(originalEntity.id, updatedEntity.id)
        assertEquals(true, updatedEntity.status)
        assertEquals(originalEntity.createdAt, updatedEntity.createdAt)
        assertNotEquals(originalEntity.updatedAt, updatedEntity.updatedAt)
    }
}
